// @jackwener/opencli/dist/src/hooks.js
/**
 * Plugin lifecycle hooks: allows plugins to tap into opencli's execution lifecycle.
 *
 * Hooks use globalThis (like the command registry) to guarantee a single shared
 * instance across all module copies — critical when TS plugins are loaded via
 * npm link / peerDependency symlinks.
 *
 * Available hooks:
 *   onStartup        — fired once after all commands & plugins are discovered
 *   onBeforeExecute  — fired before every command execution
 *   onAfterExecute   — fired after every command execution (receives result)
 */
import { log } from './logger.js';
const _hooks = globalThis.__opencli_hooks__ ??= new Map();
// ── Registration API (used by plugins) ─────────────────────────────────────
function addHook(name, fn) {
    const list = _hooks.get(name) ?? [];
    if (list.includes(fn))
        return;
    list.push(fn);
    _hooks.set(name, list);
}
/** Register a hook that fires once after all plugins are discovered. */
export function onStartup(fn) {
    addHook('onStartup', fn);
}
/** Register a hook that fires before every command execution. */
export function onBeforeExecute(fn) {
    addHook('onBeforeExecute', fn);
}
/** Register a hook that fires after every command execution with the result. */
export function onAfterExecute(fn) {
    addHook('onAfterExecute', fn);
}
// ── Emit API (used internally by opencli core) ─────────────────────────────
/**
 * Trigger all registered handlers for a hook.
 * Each handler is wrapped in try/catch — a failing hook never blocks command execution.
 */
export async function emitHook(name, ctx, result) {
    const handlers = _hooks.get(name);
    if (!handlers || handlers.length === 0)
        return;
    for (const fn of handlers) {
        try {
            await fn(ctx, result);
        }
        catch (err) {
            log.warn(`Hook ${name} handler failed: ${err instanceof Error ? err.message : String(err)}`);
        }
    }
}
/**
 * Remove all registered hooks. Intended for testing only.
 */
export function clearAllHooks() {
    _hooks.clear();
}
