// @jackwener/opencli/dist/src/browser/base-page.js
/**
 * BasePage — shared IPage method implementations for DOM helpers.
 *
 * Both Page (daemon-backed) and CDPPage (direct CDP) execute JS the same way
 * for DOM operations. This base class deduplicates ~200 lines of identical
 * click/type/scroll/wait/snapshot/interceptor methods.
 *
 * Subclasses implement the transport-specific methods: goto, evaluate,
 * getCookies, screenshot, tabs, etc.
 */
import { generateSnapshotJs, scrollToRefJs, getFormStateJs } from './dom-snapshot.js';
import { clickJs, typeTextJs, pressKeyJs, waitForTextJs, waitForCaptureJs, waitForSelectorJs, scrollJs, autoScrollJs, networkRequestsJs, waitForDomStableJs, } from './dom-helpers.js';
import { formatSnapshot } from '../snapshotFormatter.js';
export class BasePage {
    _lastUrl = null;
    /** Cached previous snapshot hashes for incremental diff marking */
    _prevSnapshotHashes = null;
    /**
     * Safely evaluate JS with pre-serialized arguments.
     * Each key in `args` becomes a `const` declaration with JSON-serialized value,
     * prepended to the JS code. Prevents injection by design.
     *
     * Usage:
     *   page.evaluateWithArgs(`(async () => { return sym; })()`, { sym: userInput })
     */
    async evaluateWithArgs(js, args) {
        const declarations = Object.entries(args)
            .map(([key, value]) => {
            if (!/^[a-zA-Z_$][a-zA-Z0-9_$]*$/.test(key)) {
                throw new Error(`evaluateWithArgs: invalid key "${key}"`);
            }
            return `const ${key} = ${JSON.stringify(value)};`;
        })
            .join('\n');
        return this.evaluate(`${declarations}\n${js}`);
    }
    // ── Shared DOM helper implementations ──
    async click(ref) {
        const result = await this.evaluate(clickJs(ref));
        // Backwards compat: old format returned 'clicked' string
        if (typeof result === 'string' || result == null)
            return;
        // JS click succeeded
        if (result.status === 'clicked')
            return;
        // JS click failed — try CDP native click if coordinates available
        if (result.x != null && result.y != null) {
            const success = await this.tryNativeClick(result.x, result.y);
            if (success)
                return;
        }
        throw new Error(`Click failed: ${result.error ?? 'JS click and CDP fallback both failed'}`);
    }
    /** Override in subclasses with CDP native click support */
    async tryNativeClick(_x, _y) {
        return false;
    }
    async typeText(ref, text) {
        await this.evaluate(typeTextJs(ref, text));
    }
    async pressKey(key) {
        await this.evaluate(pressKeyJs(key));
    }
    async scrollTo(ref) {
        return this.evaluate(scrollToRefJs(ref));
    }
    async getFormState() {
        return (await this.evaluate(getFormStateJs()));
    }
    async scroll(direction = 'down', amount = 500) {
        await this.evaluate(scrollJs(direction, amount));
    }
    async autoScroll(options) {
        const times = options?.times ?? 3;
        const delayMs = options?.delayMs ?? 2000;
        await this.evaluate(autoScrollJs(times, delayMs));
    }
    async networkRequests(includeStatic = false) {
        const result = await this.evaluate(networkRequestsJs(includeStatic));
        return Array.isArray(result) ? result : [];
    }
    async consoleMessages(_level = 'info') {
        return [];
    }
    async wait(options) {
        if (typeof options === 'number') {
            if (options >= 1) {
                try {
                    const maxMs = options * 1000;
                    await this.evaluate(waitForDomStableJs(maxMs, Math.min(500, maxMs)));
                    return;
                }
                catch {
                    // Fallback: fixed sleep
                }
            }
            await new Promise(resolve => setTimeout(resolve, options * 1000));
            return;
        }
        if (typeof options.time === 'number') {
            await new Promise(resolve => setTimeout(resolve, options.time * 1000));
            return;
        }
        if (options.selector) {
            const timeout = (options.timeout ?? 10) * 1000;
            await this.evaluate(waitForSelectorJs(options.selector, timeout));
            return;
        }
        if (options.text) {
            const timeout = (options.timeout ?? 30) * 1000;
            await this.evaluate(waitForTextJs(options.text, timeout));
        }
    }
    async snapshot(opts = {}) {
        const snapshotJs = generateSnapshotJs({
            viewportExpand: opts.viewportExpand ?? 2000,
            maxDepth: Math.max(1, Math.min(Number(opts.maxDepth) || 50, 200)),
            interactiveOnly: opts.interactive ?? false,
            maxTextLength: opts.maxTextLength ?? 120,
            includeScrollInfo: true,
            bboxDedup: true,
            previousHashes: this._prevSnapshotHashes,
        });
        try {
            const result = await this.evaluate(snapshotJs);
            // Read back the hashes stored by the snapshot for next diff
            try {
                const hashes = await this.evaluate('window.__opencli_prev_hashes');
                this._prevSnapshotHashes = typeof hashes === 'string' ? hashes : null;
            }
            catch {
                // Non-fatal: diff is best-effort
            }
            return result;
        }
        catch (err) {
            // Log snapshot failure for debugging, then fallback to basic accessibility tree
            if (process.env.DEBUG_SNAPSHOT) {
                process.stderr.write(`[snapshot] DOM snapshot failed, falling back to accessibility tree: ${err?.message?.slice(0, 200)}\n`);
            }
            return this._basicSnapshot(opts);
        }
    }
    async getCurrentUrl() {
        if (this._lastUrl)
            return this._lastUrl;
        try {
            const current = await this.evaluate('window.location.href');
            if (typeof current === 'string' && current) {
                this._lastUrl = current;
                return current;
            }
        }
        catch {
            // Best-effort
        }
        return null;
    }
    async installInterceptor(pattern) {
        const { generateInterceptorJs } = await import('../interceptor.js');
        await this.evaluate(generateInterceptorJs(JSON.stringify(pattern), {
            arrayName: '__opencli_xhr',
            patchGuard: '__opencli_interceptor_patched',
        }));
    }
    async getInterceptedRequests() {
        const { generateReadInterceptedJs } = await import('../interceptor.js');
        const result = await this.evaluate(generateReadInterceptedJs('__opencli_xhr'));
        return Array.isArray(result) ? result : [];
    }
    async waitForCapture(timeout = 10) {
        const maxMs = timeout * 1000;
        await this.evaluate(waitForCaptureJs(maxMs));
    }
    /** Fallback basic snapshot */
    async _basicSnapshot(opts = {}) {
        const maxDepth = Math.max(1, Math.min(Number(opts.maxDepth) || 50, 200));
        const code = `
      (async () => {
        function buildTree(node, depth) {
          if (depth > ${maxDepth}) return '';
          const role = node.getAttribute?.('role') || node.tagName?.toLowerCase() || 'generic';
          const name = node.getAttribute?.('aria-label') || node.getAttribute?.('alt') || node.textContent?.trim().slice(0, 80) || '';
          const isInteractive = ['a', 'button', 'input', 'select', 'textarea'].includes(node.tagName?.toLowerCase()) || node.getAttribute?.('tabindex') != null;

          ${opts.interactive ? 'if (!isInteractive && !node.children?.length) return "";' : ''}

          let indent = '  '.repeat(depth);
          let line = indent + role;
          if (name) line += ' "' + name.replace(/"/g, '\\\\\\"') + '"';
          if (node.tagName?.toLowerCase() === 'a' && node.href) line += ' [' + node.href + ']';
          if (node.tagName?.toLowerCase() === 'input') line += ' [' + (node.type || 'text') + ']';

          let result = line + '\\n';
          if (node.children) {
            for (const child of node.children) {
              result += buildTree(child, depth + 1);
            }
          }
          return result;
        }
        return buildTree(document.body, 0);
      })()
    `;
        const raw = await this.evaluate(code);
        if (opts.raw)
            return raw;
        if (typeof raw === 'string')
            return formatSnapshot(raw, opts);
        return raw;
    }
}
