// @jackwener/opencli/dist/src/browser/bridge.js
/**
 * Browser session manager — auto-spawns daemon and provides IPage.
 */
import { spawn } from 'node:child_process';
import { fileURLToPath } from 'node:url';
import * as path from 'node:path';
import * as fs from 'node:fs';
import { Page } from './page.js';
import { getDaemonHealth } from './daemon-client.js';
import { DEFAULT_DAEMON_PORT } from '../constants.js';
import { BrowserConnectError } from '../errors.js';
const DAEMON_SPAWN_TIMEOUT = 10000; // 10s to wait for daemon + extension
/**
 * Browser factory: manages daemon lifecycle and provides IPage instances.
 */
export class BrowserBridge {
    _state = 'idle';
    _page = null;
    _daemonProc = null;
    get state() {
        return this._state;
    }
    async connect(opts = {}) {
        if (this._state === 'connected' && this._page)
            return this._page;
        if (this._state === 'connecting')
            throw new Error('Already connecting');
        if (this._state === 'closing')
            throw new Error('Session is closing');
        if (this._state === 'closed')
            throw new Error('Session is closed');
        this._state = 'connecting';
        try {
            await this._ensureDaemon(opts.timeout);
            this._page = new Page(opts.workspace);
            this._state = 'connected';
            return this._page;
        }
        catch (err) {
            this._state = 'idle';
            throw err;
        }
    }
    async close() {
        if (this._state === 'closed')
            return;
        this._state = 'closing';
        // We don't kill the daemon — it's persistent.
        // Just clean up our reference.
        this._page = null;
        this._state = 'closed';
    }
    async _ensureDaemon(timeoutSeconds) {
        const effectiveSeconds = (timeoutSeconds && timeoutSeconds > 0) ? timeoutSeconds : Math.ceil(DAEMON_SPAWN_TIMEOUT / 1000);
        const timeoutMs = effectiveSeconds * 1000;
        const health = await getDaemonHealth();
        // Fast path: everything ready
        if (health.state === 'ready')
            return;
        // Daemon running but no extension — wait for extension with progress
        if (health.state === 'no-extension') {
            if (process.env.OPENCLI_VERBOSE || process.stderr.isTTY) {
                process.stderr.write('⏳ Waiting for Chrome/Chromium extension to connect...\n');
                process.stderr.write('   Make sure Chrome or Chromium is open and the OpenCLI extension is enabled.\n');
            }
            if (await this._pollUntilReady(timeoutMs))
                return;
            throw new BrowserConnectError('Browser Bridge extension not connected', 'Install the Browser Bridge:\n' +
                '  1. Download: https://github.com/jackwener/opencli/releases\n' +
                '  2. In Chrome or Chromium, open chrome://extensions → Developer Mode → Load unpacked\n' +
                '  Then run: opencli doctor', 'extension-not-connected');
        }
        // No daemon — spawn one
        const __dirname = path.dirname(fileURLToPath(import.meta.url));
        const parentDir = path.resolve(__dirname, '..');
        const daemonTs = path.join(parentDir, 'daemon.ts');
        const daemonJs = path.join(parentDir, 'daemon.js');
        const isTs = fs.existsSync(daemonTs);
        const daemonPath = isTs ? daemonTs : daemonJs;
        if (process.env.OPENCLI_VERBOSE || process.stderr.isTTY) {
            process.stderr.write('⏳ Starting daemon...\n');
        }
        const spawnArgs = isTs
            ? [process.execPath, '--import', 'tsx/esm', daemonPath]
            : [process.execPath, daemonPath];
        this._daemonProc = spawn(spawnArgs[0], spawnArgs.slice(1), {
            detached: true,
            stdio: 'ignore',
            env: { ...process.env },
        });
        this._daemonProc.unref();
        // Wait for daemon + extension
        if (await this._pollUntilReady(timeoutMs))
            return;
        const finalHealth = await getDaemonHealth();
        if (finalHealth.state === 'no-extension') {
            throw new BrowserConnectError('Browser Bridge extension not connected', 'Install the Browser Bridge:\n' +
                '  1. Download: https://github.com/jackwener/opencli/releases\n' +
                '  2. In Chrome or Chromium, open chrome://extensions → Developer Mode → Load unpacked\n' +
                '  Then run: opencli doctor', 'extension-not-connected');
        }
        throw new BrowserConnectError('Failed to start opencli daemon', `Try running manually:\n  node ${daemonPath}\nMake sure port ${DEFAULT_DAEMON_PORT} is available.`, 'daemon-not-running');
    }
    /** Poll getDaemonHealth() until state is 'ready' or deadline is reached. */
    async _pollUntilReady(timeoutMs) {
        const deadline = Date.now() + timeoutMs;
        while (Date.now() < deadline) {
            await new Promise(resolve => setTimeout(resolve, 200));
            const h = await getDaemonHealth();
            if (h.state === 'ready')
                return true;
        }
        return false;
    }
}
