// @jackwener/opencli/dist/src/browser/cdp.js
/**
 * CDP client — implements IPage by connecting directly to a Chrome/Electron CDP WebSocket.
 *
 * Fixes applied:
 * - send() now has a 30s timeout guard (P0 #4)
 * - goto() waits for Page.loadEventFired instead of hardcoded 1s sleep (P1 #3)
 * - Implemented scroll, autoScroll, screenshot, networkRequests (P1 #2)
 * - Shared DOM helper methods extracted to reduce duplication with Page (P1 #5)
 */
import { WebSocket } from 'ws';
import { request as httpRequest } from 'node:http';
import { request as httpsRequest } from 'node:https';
import { wrapForEval } from './utils.js';
import { generateStealthJs } from './stealth.js';
import { waitForDomStableJs } from './dom-helpers.js';
import { isRecord, saveBase64ToFile } from '../utils.js';
import { getAllElectronApps } from '../electron-apps.js';
import { BasePage } from './base-page.js';
const CDP_SEND_TIMEOUT = 30_000;
export class CDPBridge {
    _ws = null;
    _idCounter = 0;
    _pending = new Map();
    _eventListeners = new Map();
    async connect(opts) {
        if (this._ws)
            throw new Error('CDPBridge is already connected. Call close() before reconnecting.');
        const endpoint = opts?.cdpEndpoint ?? process.env.OPENCLI_CDP_ENDPOINT;
        if (!endpoint)
            throw new Error('CDP endpoint not provided (pass cdpEndpoint or set OPENCLI_CDP_ENDPOINT)');
        let wsUrl = endpoint;
        if (endpoint.startsWith('http')) {
            const targets = await fetchJsonDirect(`${endpoint.replace(/\/$/, '')}/json`);
            const target = selectCDPTarget(targets);
            if (!target || !target.webSocketDebuggerUrl) {
                throw new Error('No inspectable targets found at CDP endpoint');
            }
            wsUrl = target.webSocketDebuggerUrl;
        }
        return new Promise((resolve, reject) => {
            const ws = new WebSocket(wsUrl);
            const timeoutMs = (opts?.timeout ?? 10) * 1000;
            const timeout = setTimeout(() => {
                this._ws = null;
                ws.close();
                reject(new Error('CDP connect timeout'));
            }, timeoutMs);
            ws.on('open', async () => {
                clearTimeout(timeout);
                this._ws = ws;
                try {
                    await this.send('Page.enable');
                    await this.send('Page.addScriptToEvaluateOnNewDocument', { source: generateStealthJs() });
                }
                catch (err) {
                    ws.close();
                    reject(err instanceof Error ? err : new Error(String(err)));
                    return;
                }
                resolve(new CDPPage(this));
            });
            ws.on('error', (err) => {
                clearTimeout(timeout);
                reject(err);
            });
            ws.on('message', (data) => {
                try {
                    const msg = JSON.parse(data.toString());
                    if (msg.id && this._pending.has(msg.id)) {
                        const entry = this._pending.get(msg.id);
                        clearTimeout(entry.timer);
                        this._pending.delete(msg.id);
                        if (msg.error) {
                            entry.reject(new Error(msg.error.message));
                        }
                        else {
                            entry.resolve(msg.result);
                        }
                    }
                    if (msg.method) {
                        const listeners = this._eventListeners.get(msg.method);
                        if (listeners) {
                            for (const fn of listeners)
                                fn(msg.params);
                        }
                    }
                }
                catch { }
            });
        });
    }
    async close() {
        if (this._ws) {
            this._ws.close();
            this._ws = null;
        }
        for (const p of this._pending.values()) {
            clearTimeout(p.timer);
            p.reject(new Error('CDP connection closed'));
        }
        this._pending.clear();
        this._eventListeners.clear();
    }
    async send(method, params = {}, timeoutMs = CDP_SEND_TIMEOUT) {
        if (!this._ws || this._ws.readyState !== WebSocket.OPEN) {
            throw new Error('CDP connection is not open');
        }
        const id = ++this._idCounter;
        return new Promise((resolve, reject) => {
            const timer = setTimeout(() => {
                this._pending.delete(id);
                reject(new Error(`CDP command '${method}' timed out after ${timeoutMs / 1000}s`));
            }, timeoutMs);
            this._pending.set(id, { resolve, reject, timer });
            this._ws.send(JSON.stringify({ id, method, params }));
        });
    }
    on(event, handler) {
        let set = this._eventListeners.get(event);
        if (!set) {
            set = new Set();
            this._eventListeners.set(event, set);
        }
        set.add(handler);
    }
    off(event, handler) {
        this._eventListeners.get(event)?.delete(handler);
    }
    waitForEvent(event, timeoutMs = 15_000) {
        return new Promise((resolve, reject) => {
            const timer = setTimeout(() => {
                this.off(event, handler);
                reject(new Error(`Timed out waiting for CDP event '${event}'`));
            }, timeoutMs);
            const handler = (params) => {
                clearTimeout(timer);
                this.off(event, handler);
                resolve(params);
            };
            this.on(event, handler);
        });
    }
}
class CDPPage extends BasePage {
    bridge;
    _pageEnabled = false;
    // Network capture state (mirrors extension/src/cdp.ts NetworkCaptureEntry shape)
    _networkCapturing = false;
    _networkCapturePattern = '';
    _networkEntries = [];
    _pendingRequests = new Map(); // requestId → index in _networkEntries
    _pendingBodyFetches = new Set(); // track in-flight getResponseBody calls
    _consoleMessages = [];
    _consoleCapturing = false;
    constructor(bridge) {
        super();
        this.bridge = bridge;
    }
    async goto(url, options) {
        if (!this._pageEnabled) {
            await this.bridge.send('Page.enable');
            this._pageEnabled = true;
        }
        const loadPromise = this.bridge.waitForEvent('Page.loadEventFired', 30_000).catch(() => { });
        await this.bridge.send('Page.navigate', { url });
        await loadPromise;
        this._lastUrl = url;
        if (options?.waitUntil !== 'none') {
            const maxMs = options?.settleMs ?? 1000;
            await this.evaluate(waitForDomStableJs(maxMs, Math.min(500, maxMs)));
        }
    }
    async evaluate(js) {
        const expression = wrapForEval(js);
        const result = await this.bridge.send('Runtime.evaluate', {
            expression,
            returnByValue: true,
            awaitPromise: true,
        });
        if (result.exceptionDetails) {
            throw new Error('Evaluate error: ' + (result.exceptionDetails.exception?.description || 'Unknown exception'));
        }
        return result.result?.value;
    }
    async getCookies(opts = {}) {
        const result = await this.bridge.send('Network.getCookies', opts.url ? { urls: [opts.url] } : {});
        const cookies = isRecord(result) && Array.isArray(result.cookies) ? result.cookies : [];
        const domain = opts.domain;
        return domain
            ? cookies.filter((cookie) => isCookie(cookie) && matchesCookieDomain(cookie.domain, domain))
            : cookies;
    }
    async screenshot(options = {}) {
        const result = await this.bridge.send('Page.captureScreenshot', {
            format: options.format ?? 'png',
            quality: options.format === 'jpeg' ? (options.quality ?? 80) : undefined,
            captureBeyondViewport: options.fullPage ?? false,
        });
        const base64 = isRecord(result) && typeof result.data === 'string' ? result.data : '';
        if (options.path) {
            await saveBase64ToFile(base64, options.path);
        }
        return base64;
    }
    async startNetworkCapture(pattern = '') {
        // Always update the filter pattern
        this._networkCapturePattern = pattern;
        // Reset state only on first start; avoid wiping entries if already capturing
        if (!this._networkCapturing) {
            this._networkEntries = [];
            this._pendingRequests.clear();
            this._pendingBodyFetches.clear();
            await this.bridge.send('Network.enable');
            // Step 1: Record request method/url on requestWillBeSent
            this.bridge.on('Network.requestWillBeSent', (params) => {
                const p = params;
                if (!this._networkCapturePattern || p.request.url.includes(this._networkCapturePattern)) {
                    const idx = this._networkEntries.push({
                        url: p.request.url,
                        method: p.request.method,
                        timestamp: p.timestamp,
                    }) - 1;
                    this._pendingRequests.set(p.requestId, idx);
                }
            });
            // Step 2: Fill in response metadata on responseReceived
            this.bridge.on('Network.responseReceived', (params) => {
                const p = params;
                const idx = this._pendingRequests.get(p.requestId);
                if (idx !== undefined) {
                    this._networkEntries[idx].responseStatus = p.response.status;
                    this._networkEntries[idx].responseContentType = p.response.mimeType || '';
                }
            });
            // Step 3: Fetch body on loadingFinished (body is only reliably available after this)
            this.bridge.on('Network.loadingFinished', (params) => {
                const p = params;
                const idx = this._pendingRequests.get(p.requestId);
                if (idx !== undefined) {
                    const bodyFetch = this.bridge.send('Network.getResponseBody', { requestId: p.requestId }).then((result) => {
                        const r = result;
                        if (typeof r?.body === 'string') {
                            this._networkEntries[idx].responsePreview = r.base64Encoded
                                ? `base64:${r.body.slice(0, 4000)}`
                                : r.body.slice(0, 4000);
                        }
                    }).catch(() => {
                        // Body unavailable for some requests (e.g. uploads) — non-fatal
                    }).finally(() => {
                        this._pendingBodyFetches.delete(bodyFetch);
                    });
                    this._pendingBodyFetches.add(bodyFetch);
                    this._pendingRequests.delete(p.requestId);
                }
            });
            this._networkCapturing = true;
        }
    }
    async readNetworkCapture() {
        // Await all in-flight body fetches so entries have responsePreview populated
        if (this._pendingBodyFetches.size > 0) {
            await Promise.all([...this._pendingBodyFetches]);
        }
        const entries = [...this._networkEntries];
        this._networkEntries = [];
        return entries;
    }
    async consoleMessages(level = 'all') {
        if (!this._consoleCapturing) {
            await this.bridge.send('Runtime.enable');
            this.bridge.on('Runtime.consoleAPICalled', (params) => {
                const p = params;
                const text = (p.args || []).map(a => a.value !== undefined ? String(a.value) : (a.description || '')).join(' ');
                this._consoleMessages.push({ type: p.type, text, timestamp: p.timestamp });
                if (this._consoleMessages.length > 500)
                    this._consoleMessages.shift();
            });
            // Capture uncaught exceptions as error-level messages
            this.bridge.on('Runtime.exceptionThrown', (params) => {
                const p = params;
                const desc = p.exceptionDetails?.exception?.description || p.exceptionDetails?.text || 'Unknown exception';
                this._consoleMessages.push({ type: 'error', text: desc, timestamp: p.timestamp });
                if (this._consoleMessages.length > 500)
                    this._consoleMessages.shift();
            });
            this._consoleCapturing = true;
        }
        if (level === 'all')
            return [...this._consoleMessages];
        // 'error' level includes both console.error() and uncaught exceptions
        if (level === 'error')
            return this._consoleMessages.filter(m => m.type === 'error' || m.type === 'warning');
        return this._consoleMessages.filter(m => m.type === level);
    }
    async tabs() {
        return [];
    }
    async selectTab(_index) {
        // Not supported in direct CDP mode
    }
}
function isCookie(value) {
    return isRecord(value)
        && typeof value.name === 'string'
        && typeof value.value === 'string'
        && typeof value.domain === 'string';
}
function matchesCookieDomain(cookieDomain, targetDomain) {
    const normalizedCookieDomain = cookieDomain.replace(/^\./, '').toLowerCase();
    const normalizedTargetDomain = targetDomain.replace(/^\./, '').toLowerCase();
    return normalizedTargetDomain === normalizedCookieDomain
        || normalizedTargetDomain.endsWith(`.${normalizedCookieDomain}`);
}
function selectCDPTarget(targets) {
    const preferredPattern = compilePreferredPattern(process.env.OPENCLI_CDP_TARGET);
    const ranked = targets
        .map((target, index) => ({ target, index, score: scoreCDPTarget(target, preferredPattern) }))
        .filter(({ score }) => Number.isFinite(score))
        .sort((a, b) => {
        if (b.score !== a.score)
            return b.score - a.score;
        return a.index - b.index;
    });
    return ranked[0]?.target;
}
function scoreCDPTarget(target, preferredPattern) {
    if (!target.webSocketDebuggerUrl)
        return Number.NEGATIVE_INFINITY;
    const type = (target.type ?? '').toLowerCase();
    const url = (target.url ?? '').toLowerCase();
    const title = (target.title ?? '').toLowerCase();
    const haystack = `${title} ${url}`;
    if (!haystack.trim() && !type)
        return Number.NEGATIVE_INFINITY;
    if (haystack.includes('devtools'))
        return Number.NEGATIVE_INFINITY;
    if (type === 'background_page' || type === 'service_worker')
        return Number.NEGATIVE_INFINITY;
    let score = 0;
    if (preferredPattern && preferredPattern.test(haystack))
        score += 1000;
    if (type === 'app')
        score += 120;
    else if (type === 'webview')
        score += 100;
    else if (type === 'page')
        score += 80;
    else if (type === 'iframe')
        score += 20;
    if (url.startsWith('http://localhost') || url.startsWith('https://localhost'))
        score += 90;
    if (url.startsWith('file://'))
        score += 60;
    if (url.startsWith('http://127.0.0.1') || url.startsWith('https://127.0.0.1'))
        score += 50;
    if (url.startsWith('about:blank'))
        score -= 120;
    if (url === '' || url === 'about:blank')
        score -= 40;
    if (title && title !== 'devtools')
        score += 25;
    // Boost score for known Electron app names from the registry (builtin + user-defined)
    const appNames = Object.values(getAllElectronApps()).map(a => (a.displayName ?? a.processName).toLowerCase());
    for (const name of appNames) {
        if (title.includes(name)) {
            score += 120;
            break;
        }
    }
    for (const name of appNames) {
        if (url.includes(name)) {
            score += 100;
            break;
        }
    }
    return score;
}
function compilePreferredPattern(raw) {
    const value = raw?.trim();
    if (!value)
        return undefined;
    return new RegExp(escapeRegExp(value.toLowerCase()));
}
function escapeRegExp(value) {
    return value.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}
export const __test__ = {
    selectCDPTarget,
    scoreCDPTarget,
};
function fetchJsonDirect(url) {
    return new Promise((resolve, reject) => {
        const parsed = new URL(url);
        const request = (parsed.protocol === 'https:' ? httpsRequest : httpRequest)(parsed, (res) => {
            const statusCode = res.statusCode ?? 0;
            if (statusCode < 200 || statusCode >= 300) {
                res.resume();
                reject(new Error(`Failed to fetch CDP targets: HTTP ${statusCode}`));
                return;
            }
            const chunks = [];
            res.on('data', (chunk) => chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk)));
            res.on('end', () => {
                try {
                    resolve(JSON.parse(Buffer.concat(chunks).toString('utf8')));
                }
                catch (error) {
                    reject(error instanceof Error ? error : new Error(String(error)));
                }
            });
        });
        request.on('error', reject);
        request.setTimeout(10_000, () => request.destroy(new Error('Timed out fetching CDP targets')));
        request.end();
    });
}
